components/ai_chat/core/browser/associated_content_driver.h

```diff
@@ -0,0 +1,167 @@
+  // will not be detected by the on-demand techniques used by GetPageContent.
+  // For example for sites where GetPageContent does not read the live DOM but
+  // reads static JS from HTML that doesn't change for same-page navigation and
+  // we need to intercept new JS data from subresource loads.
+  void OnPageContentUpdated(std::string content,
+                            bool is_video,
+                            std::string invalidation_token);
+
+  // To be called when a page navigation is detected and a new conversation
+  // is expected.
+  void OnNewPage(int64_t navigation_id) override;
+
+  // Begin the alternative content fetching (print preview / OCR) by
+  // sending a message to an observer with access to the layer this can
+  // be performed on (i.e. the UI layer can use Print Preview).
+  // Since this AssociatedContentDriver might be open in multiple WebUIs,
+  // we can no longer rely on the WebUI class observing its conversation 1:1.
+  // We might end up with multiple UIs performing the print preview fetching
+  // for a single conversation. Instead we need some single entity that is
+  // able to perform this work, either by relocating AIChatTabHelper to
+  // brave/browser/ui, or having a singleton/KeyedService observe this event
+  // void NotifyPrintPreviewRequested(bool is_pdf);
+ private:
+  friend class ::AIChatUIBrowserTest;
+  FRIEND_TEST_ALL_PREFIXES(::AIChatUIBrowserTest, PrintPreviewFallback);
+  FRIEND_TEST_ALL_PREFIXES(AssociatedContentDriverUnitTest,
+                           ParseSearchQuerySummaryResponse);
+
+  void OnGeneratePageContentComplete(int64_t navigation_id,
+                                     std::string contents_text,
+                                     bool is_video,
+                                     std::string invalidation_token);
+  void OnExistingGeneratePageContentComplete(
+      ConversationHandler::GetPageContentCallback callback,
+      int64_t navigation_id);
+
+  void OnSearchSummarizerKeyFetched(
+      ConversationHandler::GetStagedEntriesCallback callback,
+      int64_t navigation_id,
+      const std::optional<std::string>& key);
+  void OnSearchQuerySummaryFetched(
+      ConversationHandler::GetStagedEntriesCallback callback,
+      int64_t navigation_id,
+      api_request_helper::APIRequestResult result);
+  static std::optional<SearchQuerySummary> ParseSearchQuerySummaryResponse(
+      const base::Value& value);
+
+  raw_ptr<PrefService> pref_service_;
+  raw_ptr<AIChatMetrics> ai_chat_metrics_;
+  std::unique_ptr<AIChatCredentialManager> credential_manager_;
+
+  scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory_;
+
+  // Used for fetching search query summary.
+  std::unique_ptr<api_request_helper::APIRequestHelper> api_request_helper_;
+
+  base::ObserverList<Observer> observers_;
+  std::unique_ptr<base::OneShotEvent> on_page_text_fetch_complete_;
+  bool is_page_text_fetch_in_progress_ = false;
```

Comment:
I think rather than having two separate variables here we could only set `on_page_text_fetch_complete_` while we're fetching - then the fetch in progress check becomes

```cpp
if (on_page_text_fetch_complete_)
```

Reply:
I did that - thanks for the suggestion b881722559bc5a51dd097570d9f1bd0522e2a4e2